docs/tutorials/forum-application/3.send-message.md

```diff
@@ -0,0 +1,381 @@
+			respTxs[i] = &abci.ExecTxResult{Code: CodeTypeEncodingError}
+		} else {
+			// Check if this sender already existed; if not, add the user too
+			err := UpdateOrSetUser(app.state.DB, msg.Sender, false, app.onGoingBlock)
+			if err != nil {
+				panic(err)
+			}
+			// Add the message for this sender
+			message, err := model.AppendToExistingMsgs(app.state.DB, *msg)
+			if err != nil {
+				panic(err)
+			}
+			app.onGoingBlock.Set([]byte(msg.Sender+"msg"), []byte(message))
+			chatHistory, err := model.AppendToChat(app.state.DB, *msg)
+			if err != nil {
+				panic(err)
+			}
+			// Append messages to chat history
+			app.onGoingBlock.Set([]byte("history"), []byte(chatHistory))
+			// This adds the user to the DB, but the data is not committed nor persisted until Commit is called
+			respTxs[i] = &abci.ExecTxResult{Code: abci.CodeTypeOK}
+			app.state.Size++
+		}
+	}
+	app.state.Height = req.Height
+
+	response := &abci.FinalizeBlockResponse{TxResults: respTxs, AppHash: app.state.Hash()}
+	return response, nil
+}
+```
+
+**Explanation of code:**
+
+`FinalizeBlock` function iterates over the transactions (`req.Txs`) in the block. If a transaction is a ban transaction (`isBanTx(tx)`),
+it updates or sets a user in the application's database and assigns a response code. If it is not a ban transaction, the loop breaks.
+
+After the ban transactions, the method parses the remaining transactions using `model.ParseMessage(tx)`. It updates or sets
+the sender as a user in the database, appends the message to the existing messages, appends the message to the chat history,
+and assigns a response code. Finally, it updates the blockchain state and returns the response object.
+
+*Tip: The function `FinalizeBlock` finalizes processing of a tx. All the state change that happen in this function are finalized.
+However, they are not yet persisted to database. This is done in next step, i.e. `Commit`*
+
+Note: You will learn about different packages and functions like `model.AppendToExistingMsgs` in the upcoming sections.
+In this section you will learn about `app.go` file only.
+
+## Commit
+
+The `Commit` method is used to persist the changes made to the application state during the FinalizeBlock method.
+
+After calling `FinalizeBlock`, the state changes are finalized but not yet persisted to the database. The `Commit` method
+is responsible for persisting these changes to the database, ensuring that they are durable and can be retrieved later.
+
+Typically, the `Commit` method updates the application's blockchain state by saving the modified data to a persistent
+storage system, such as a database or a file. This allows the application to maintain a consistent and reliable state
+across different blocks in the blockchain.
+
+The Commit method takes in a `CommitRequest` object and returns a `CommitResponse` object, which can include information
+about the `commit` process, such as the new state hash or any errors that occurred during the commit.
```

Comment:
The hash has been moved to `FinalizeBlock` in `v0.38.x`

Reply:
Thanks, removed that information. 

By the way, noticed the `/spec/core/state` doc has some old data for CommitResponse, I've added that into an [issue](https://github.com/cometbft/cometbft/issues/2834) I had opened to fix this page